### ISSUE TITLE
no-extra-semi false positive with `ExportNamedDeclaration` & `ExportDefaultDeclaration`

### ISSUE THREAD

**Issue by mathiasbynens** [eslint/eslint#4915]
The semicolons in these examples are incorrectly flagged as unnecessary:

```js
// ExportNamedDeclaration
export const x = 42;
```

```js
// ExportDefaultDeclaration
export default 42;
```

**Comment by eslintbot** [comment]
@mathiasbynens Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
No, those are unnecessary due to ASI. Perhaps you are thinking about the case when certain expressions follow?

**Comment by platinumazure** [comment]
no-extra-semi is not supposed to warn on ASI locations (that's the job of the semi rule). This rule is supposed to warn on things like function declarations (which are unambiguously parsed as statements even without a semicolon in most cases), in other words things that shouldn't need a semicolon ever. I think exports require semicolons (or ASI) but I'm not sure.

**Comment by michaelficarra** [comment]
Ah, I see. So no-extra-semi literally just flags `EmptyStatement`s? Yes, the export declarations will consume that semicolon, so there should be no `EmptyStatement`s here and thus no report.

**Comment by btmills** [comment]
@mathiasbynens what version of ESLint and what parser are you using? Both of those examples pass as test cases added off of `master` and `v1.10.3`.

**Comment by mathiasbynens** [comment]
@btmills I noticed this while submitting a patch to [babel/babel](https://github.com/babel/babel) and triggering the CI build. They seem to be using eslint v1.8.0.

If this has been fixed in the mean time, all the better!

**Comment by nzakas** [comment]
Confirmed previously fixed.

**Comment by platinumazure** [comment]
@nzakas I'm not seeing any unit tests in no-extra-semi around `export` statements. Were they covered in some other test case?

**Comment by nzakas** [comment]
Actually, `no-extra-semi`

**Comment by nzakas** [comment]
Oops, premature Enter. `no-extra-semi` has never flagged this case, so I'm curious if that's not the rule in question.

@mathiasbynens please paste the actual output you're seeing (note we ask for this in the first comment).

**Comment by mathiasbynens** [comment]
@nzakas: https://travis-ci.org/babel/babel/jobs/101628941#L116

**Comment by nzakas** [comment]
@mathiasbynens we only ask for information that's helpful. :) Sometimes people misread or misrepresent an actual error, that's why we ask for the complete console output. It really helps cut down on the amount of time it takes to triage a bug.

So this is definitely in `no-extra-semi`, but I can't reproduce. I also can't see how the rule ever would flag such code based on the implementation. Per @platinumazure's comment, I'll add some tests to be sure, but pretty certain this is most likely an issue on Babel's side (possibly with the way it's treating code before passing to ESLint).

**Comment by mathiasbynens** [comment]
Ref. https://github.com/tc39/ecma262/issues/284

**Comment by mathiasbynens** [comment]
More tests (https://github.com/tc39/ecma262/issues/284):

> However, [the `ExportDeclaration` grammar](https://tc39.github.io/ecma262/#prod-ExportDeclaration) doesn’t include the semicolon for the following cases:
> 
> * `export HoistableDeclaration`, e.g. `export function foo() {}`
> * `export ClassDeclaration`, e.g. `export class Bar {}`
> * `export default HoistableDeclaration`, e.g. `export default function baz() {}`
> * `export default ClassDeclaration`, e.g. `export default class Qux {}`
